motulator/grid/model/_grid_volt_source.py

```diff
@@ -0,0 +1,372 @@
+    phi : callable, optional
+        Phase shift of the grid voltage (rad). Default is 0.
+    e_g_neg_abs : float | callable, optional
+        Negative sequence voltage magnitude, phase-to-ground peak value (V).
+        Default is None.
+    phi_neg : float, optional
+        Phase shift of the negative sequence component (rad). Default is 0.
+    w_g : callable, optional
+        Grid frequency (rad/s) as a function of time, `w_g(t)`. If given, w_g
+        will be used to compute grid voltage angle instead of w_gN. Default
+        is None.
+    e_ga_abs : callable, optional
+        Phase A voltage magnitude, phase-to-ground peak value (V). If given,
+        this will be used instead of e_g_abs. Default is None.
+    e_gb_abs : callable, optional
+        Phase B voltage magnitude, phase-to-ground peak value (V). Default is
+        None.
+    e_gc_abs : callable, optional
+        Phase C voltage magnitude, phase-to-ground peak value (V). Default is
+        None.
+
+    """
+
+    def __init__(
+            self,
+            w_gN,
+            e_g_abs,
+            phi=lambda t: 0,
+            e_g_neg_abs=0,
+            phi_neg=0,
+            w_g=None,
+            e_ga_abs=None,
+            e_gb_abs=None,
+            e_gc_abs=None):
+        super().__init__()
+        self.par = SimpleNamespace(
+            w_gN=w_gN,
+            e_g_abs=e_g_abs,
+            phi=phi,
+            e_g_neg_abs=e_g_neg_abs,
+            phi_neg=phi_neg,
+            w_g=w_g,
+            e_ga_abs=e_ga_abs,
+            e_gb_abs=e_gb_abs,
+            e_gc_abs=e_gc_abs,
+        )
+        # states
+        self.state = SimpleNamespace(exp_j_theta_p=complex(1))
+        # Store the solutions in these lists
+        self.sol_states = SimpleNamespace(exp_j_theta_p=[])
+
+    def voltages(self, t, theta_p):
+        """
+        Compute the grid voltage in stationary frame.
+        
+        Parameters
+        ----------
+        t : float
+            Time (s).
+        theta_p : float
+            Phase A voltage angle (rad).
+
+        Returns
+        -------
+        e_gs : complex
+            Grid complex voltage (V).
+
+        """
+        # Phase voltage magnitudes
+        par = self.par
+        e_g_abs = par.e_g_abs(t) if callable(par.e_g_abs) else par.e_g_abs
+        e_ga_abs = par.e_ga_abs(t) if par.e_ga_abs is not None else e_g_abs
+        e_gb_abs = par.e_gb_abs(t) if par.e_gb_abs is not None else e_g_abs
+        e_gc_abs = par.e_gc_abs(t) if par.e_gc_abs is not None else e_g_abs
+
+        # Calculation of the three-phase voltages
+        e_ga = e_ga_abs*np.cos(theta_p - par.phi(t))
+        e_gb = e_gb_abs*np.cos(theta_p - 2*np.pi/3 - par.phi(t))
+        e_gc = e_gc_abs*np.cos(theta_p - 4*np.pi/3 - par.phi(t))
+
+        e_gs = abc2complex([e_ga, e_gb, e_gc])
+
+        # Adding negative sequence component if given
+        if par.e_g_neg_abs is not None:
```

Comment:
Yes, it might be cleared to delete the option to give the phase voltages magnitudes here.

Could we also delete w_gN and rely on w_g only?

We may have a separate class (or some other approach) in which the three space voltages can be specified as arbitrary functions of time. This could be implemented later.